`Manifest.from(Object...)` does not support merging Manifest implementations

<!--- 
Please follow the instructions below. 
We receive dozens of issues every week, so to stay productive, we will close issues that don't provide enough information. 

Please open Android-related issues on the Android Issue Tracker at https://source.android.com/source/report-bugs
Please open Gradle Native-related issues at https://github.com/gradle/gradle-native/issues
-->

<!--- Provide a brief summary of the issue in the title above -->

### Expected Behavior
<!--- Tell us what should happen -->
One should be able to pass an implementation of `org.gradle.api.java.archives.Manifest` to `Manifest.from(Object...)` to get the argument manifest merged into the receiver manifest.

The [javadoc](https://github.com/gradle/gradle/blob/270f942676177479a8e389cdd734a18771bc4982/subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/Manifest.java#L84-L85) indicates this is supported. 

### Current Behavior
<!--- Tell us what happens instead of the expected behavior -->
The current implementation of `DefaultManifestMergeSpec` requires the manifest object to be an instance of the internal `DefaultManifest` type:

https://github.com/gradle/gradle/blob/270f942676177479a8e389cdd734a18771bc4982/subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultManifestMergeSpec.java#L152

rather than the `Manifest` type. When the manifest object is not an instance of the internal `DefaultManifest` type, the code attempts to resolve the manifest object as a file path which fails.

### Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

The is affecting the use of custom manifest object in plugins as they cannot merge a manifest object unless it is of the internal `DefaultManifest` type.

### Steps to Reproduce 
<!---
Provide a self-contained example project as an attached archive or a Github project.
You can use the following template with a Gradle GitHub action set up to showcase your problem: https://github.com/gradle/gradle-issue-reproducer
In the rare cases where this is infeasible, we will also accept a detailed set of instructions.
-->
This can be seen when using the latest fix https://github.com/bndtools/bnd/pull/5276 to the Bnd Gradle plugin (6.4.0-SNAPSHOT) when building JUnit 5 which also uses the `com.github.johnrengelman.shadow` plugin. These two plugins interact over the jar task's manifest object. With the Bnd fix, the following exception occurs.

```
Caused by: org.gradle.internal.typeconversion.UnsupportedNotationException: Cannot convert the provided notation to a File or URI: org.gradle.api.java.archives.internal.DefaultManifest@1ab3a446.
The following types/formats are supported:
  - A String or CharSequence path, for example 'src/main/java' or '/usr/include'.
  - A String or CharSequence URI, for example 'file:/usr/include'.
  - A File instance.
  - A Path instance.
  - A Directory instance.
  - A RegularFile instance.
  - A URI or URL instance.
  - A TextResource instance.
        at org.gradle.internal.typeconversion.ErrorHandlingNotationParser.parseNotation(ErrorHandlingNotationParser.java:57)
        at org.gradle.api.internal.file.AbstractFileResolver.convertObjectToFile(AbstractFileResolver.java:105)
        at org.gradle.api.internal.file.AbstractBaseDirFileResolver.doResolve(AbstractBaseDirFileResolver.java:63)
        at org.gradle.api.internal.file.AbstractFileResolver.resolve(AbstractFileResolver.java:74)
        at org.gradle.api.internal.file.AbstractFileResolver.resolve(AbstractFileResolver.java:48)
        at org.gradle.api.java.archives.internal.DefaultManifest.read(DefaultManifest.java:281)
        at org.gradle.api.java.archives.internal.DefaultManifest.<init>(DefaultManifest.java:77)
        at org.gradle.api.java.archives.internal.DefaultManifestMergeSpec.createManifest(DefaultManifestMergeSpec.java:155)
        at org.gradle.api.java.archives.internal.DefaultManifestMergeSpec.merge(DefaultManifestMergeSpec.java:84)
        at org.gradle.api.java.archives.internal.DefaultManifestMergeSpec$merge.call(Unknown Source)
        at com.github.jengelman.gradle.plugins.shadow.tasks.DefaultInheritManifest$_getEffectiveManifest_closure1.doCall(DefaultInheritManifest.groovy:65)

```
I have a PR I will submit with the fix for this issue.
 
### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL:
https://scans.gradle.com/s/huhyrwmnaz73w

--- a/subprojects/docs/src/docs/release/notes.md
+++ b/subprojects/docs/src/docs/release/notes.md
@@ -9,6 +9,7 @@ Include only their name, impactful features should be called out separately belo
  THiS LIST SHOULD BE ALPHABETIZED BY [PERSON NAME] - the docs:updateContributorsInReleaseNotes task will enforce this ordering, which is case-insensitive.
 -->
 We would like to thank the following community members for their contributions to this release of Gradle:
+[BJ Hargrave](https://github.com/bjhargrave),
 [altrisi](https://github.com/altrisi),
 [aSemy](https://github.com/aSemy),
 [Ashwin Pankaj](https://github.com/ashwinpankaj),

--- a/subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/CustomManifestInternalWrapper.java
+++ b/subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/CustomManifestInternalWrapper.java
@@ -62,12 +62,14 @@ public class CustomManifestInternalWrapper implements ManifestInternal {
 
     @Override
     public Manifest attributes(Map<String, ?> attributes) throws ManifestException {
-        return delegate.attributes(attributes);
+        delegate.attributes(attributes);
+        return this;
     }
 
     @Override
     public Manifest attributes(Map<String, ?> attributes, String sectionName) throws ManifestException {
-        return delegate.attributes(attributes, sectionName);
+        delegate.attributes(attributes, sectionName);
+        return this;
     }
 
     @Override
@@ -77,21 +79,25 @@ public class CustomManifestInternalWrapper implements ManifestInternal {
 
     @Override
     public Manifest writeTo(Object path) {
-        return delegate.writeTo(path);
+        delegate.writeTo(path);
+        return this;
     }
 
     @Override
     public Manifest from(Object... mergePath) {
-        return delegate.from(mergePath);
+        delegate.from(mergePath);
+        return this;
     }
 
     @Override
     public Manifest from(Object mergePath, Closure<?> closure) {
-        return delegate.from(mergePath, closure);
+        delegate.from(mergePath, closure);
+        return this;
     }
 
     @Override
     public Manifest from(Object mergePath, Action<ManifestMergeSpec> action) {
-        return delegate.from(mergePath, action);
+        delegate.from(mergePath, action);
+        return this;
     }
 }

--- a/subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultManifestMergeSpec.java
+++ b/subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultManifestMergeSpec.java
@@ -81,13 +81,13 @@ public class DefaultManifestMergeSpec implements ManifestMergeSpec {
         mergedManifest.getAttributes().putAll(baseManifest.getAttributes());
         mergedManifest.getSections().putAll(baseManifest.getSections());
         for (Object mergePath : mergePaths) {
-            DefaultManifest manifestToMerge = createManifest(mergePath, fileResolver, contentCharset);
+            Manifest manifestToMerge = createManifest(mergePath, fileResolver, contentCharset);
             mergedManifest = mergeManifest(mergedManifest, manifestToMerge, fileResolver);
         }
         return mergedManifest;
     }
 
-    private DefaultManifest mergeManifest(DefaultManifest baseManifest, DefaultManifest toMergeManifest, PathToFileResolver fileResolver) {
+    private DefaultManifest mergeManifest(Manifest baseManifest, Manifest toMergeManifest, PathToFileResolver fileResolver) {
         DefaultManifest mergedManifest = new DefaultManifest(fileResolver);
         mergeSection(null, mergedManifest, baseManifest.getAttributes(), toMergeManifest.getAttributes());
         Set<String> allSections = Sets.union(baseManifest.getSections().keySet(), toMergeManifest.getSections().keySet());
@@ -99,7 +99,7 @@ public class DefaultManifestMergeSpec implements ManifestMergeSpec {
         return mergedManifest;
     }
 
-    private void mergeSection(String section, DefaultManifest mergedManifest, Attributes baseAttributes, Attributes mergeAttributes) {
+    private void mergeSection(String section, Manifest mergedManifest, Attributes baseAttributes, Attributes mergeAttributes) {
         Map<String, Object> mergeOnlyAttributes = new LinkedHashMap<String, Object>(mergeAttributes);
         Set<DefaultManifestMergeDetails> mergeDetailsSet = new LinkedHashSet<DefaultManifestMergeDetails>();
 
@@ -138,7 +138,7 @@ public class DefaultManifestMergeSpec implements ManifestMergeSpec {
         }
     }
 
-    private void addMergeDetailToManifest(String section, DefaultManifest mergedManifest, DefaultManifestMergeDetails mergeDetails) {
+    private void addMergeDetailToManifest(String section, Manifest mergedManifest, DefaultManifestMergeDetails mergeDetails) {
         if (!mergeDetails.isExcluded()) {
             if (section == null) {
                 mergedManifest.attributes(WrapUtil.toMap(mergeDetails.getKey(), mergeDetails.getValue()));
@@ -148,9 +148,9 @@ public class DefaultManifestMergeSpec implements ManifestMergeSpec {
         }
     }
 
-    private DefaultManifest createManifest(Object mergePath, PathToFileResolver fileResolver, String contentCharset) {
-        if (mergePath instanceof DefaultManifest) {
-            return ((DefaultManifest) mergePath).getEffectiveManifest();
+    private Manifest createManifest(Object mergePath, PathToFileResolver fileResolver, String contentCharset) {
+        if (mergePath instanceof Manifest) {
+            return ((Manifest) mergePath).getEffectiveManifest();
         }
         return new DefaultManifest(mergePath, fileResolver, contentCharset);
     }

--- a/subprojects/platform-jvm/src/test/groovy/org/gradle/api/java/archives/internal/DefaultManifestMergeSpecTest.groovy
+++ b/subprojects/platform-jvm/src/test/groovy/org/gradle/api/java/archives/internal/DefaultManifestMergeSpecTest.groovy
@@ -17,6 +17,7 @@ package org.gradle.api.java.archives.internal
 
 import org.gradle.api.Action
 import org.gradle.api.internal.file.FileResolver
+import org.gradle.api.java.archives.Manifest
 import org.gradle.api.java.archives.ManifestMergeDetails
 import org.gradle.test.fixtures.file.TestFile
 import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
@@ -124,6 +125,16 @@ class DefaultManifestMergeSpecTest extends Specification {
 
         expect:
         mergeSpec.merge(baseManifest, Mock(FileResolver)).sections.section == [key2: 'mergeValue2']
+    }
+
+    def mergeWithForeignManifest() {
+        DefaultManifest baseManifest = new DefaultManifest(Mock(FileResolver))
+        baseManifest.attributes key1: 'value1'
+        Manifest foreign = new CustomManifestInternalWrapper(new DefaultManifest(Mock(FileResolver)).attributes( key2: 'value2'))
+        mergeSpec.from(foreign)
+
+        expect:
+        mergeSpec.merge(baseManifest, Mock(FileResolver)).attributes == [key1: 'value1', key2: 'value2'] + MANIFEST_VERSION_MAP
     }
 
     def mergeShouldWinByDefault() {

